### PR TITLE
Ensure explicitly implemented members are discoverable in BeEquivalentTo

### LINKED ISSUE #2146: Explicitly implemented properties on the subject are not found during object graph comparison
```cs
interface IPerson
{
    string Name { get }
}

class Person : IPerson
{
    string IPerson.Name => "Bob";
}

IPerson person = new Person();

person.Should().BeEquivalentTo(new { Name = "Bob" });
```

Expected: Should not throw 
Actual: Throws with message "Expectation has property person.Name that the other object does not have."

### PR DESCRIPTION
**Pull request by dennisdoomen**
* When trying to find a property on the subject to match the one on the expectation, if a normal property exists next to an implicitly implemented one, the normal property will be selected.
* Some tests related to explicit interface implementations assumed a dependency on whether or not `RespectingDeclaredTypes` and `RespectingRuntimeTypes` was used. But those options don't affect the subject.
* The `WithMapping` construct also didn't support mapping to properties implemented through explicitly-implemented interfaces.
* Removed the "non-private" part names since visibility is now used to find members

**Note**   I had to extend `MemberVisibility` to optionally include explicitly implemented members. This _could_ be perceived as a breaking change. But it won't affect consumers, so for now, I treat it as non-breaking.

**Review**
Review by commit and with whitespace changes hidden

Fixes #2146 

## IMPORTANT 

* [x] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [x] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [x] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [x] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

### CODE REVIEW

File: Src/FluentAssertions/Common/TypeMemberReflector.cs [hunk 1]
```diff
@@ -148,9 +157,17 @@ private static IEnumerable<FieldInfo> GetFieldsFromHierarchy(Type typeToReflect,
         return members;
     }
 
-    private static bool HasNonPrivateGetter(PropertyInfo propertyInfo)
+    private bool HasNonPrivateGetter(PropertyInfo propertyInfo)
```
Comment (jnyrup) [inline Src/FluentAssertions/Common/TypeMemberReflector.cs:160]:
```suggestion
    private static bool HasNonPrivateGetter(PropertyInfo propertyInfo, MemberVisibility visibility)
```

File: Src/FluentAssertions/Common/TypeMemberReflector.cs [hunk 2]
```diff
@@ -40,19 +40,23 @@ where HasNonPrivateGetter(propertyInfo)
...
         });
     }
 
-    private static FieldInfo[] LoadNonPrivateFields(Type typeToReflect, MemberVisibility visibility)
+    private static bool IsInternal(PropertyInfo property)
+    {
+        return property.GetMethod is { IsAssembly: true } or { IsFamilyOrAssembly: true };
+    }
+
+    private void LoadNonPrivateFields()
```
Comment (jnyrup) [inline Src/FluentAssertions/Common/TypeMemberReflector.cs:59]:
```suggestion
    private static FieldInfo[] LoadNonPrivateFields(Type typeToReflect, MemberVisibility visibility)
```

File: Src/FluentAssertions/Common/TypeExtensions.cs [hunk 3]
```diff
@@ -180,19 +180,12 @@ public static bool OverridesEquals(this Type type)
...
-                return property;
-            }
-
-            type = type.BaseType;
-        }
+        var properties = type.GetNonPrivateProperties(visibility);
 
-        return null;
+        return properties.FirstOrDefault(p =>
+            p.Name == propertyName || p.Name.EndsWith("." + propertyName, StringComparison.OrdinalIgnoreCase));
```
Comment (jnyrup) [inline Src/FluentAssertions/Common/TypeExtensions.cs:188]:
```suggestion
        return Array.Find(properties, p =>
            p.Name == propertyName || p.Name.EndsWith("." + propertyName, StringComparison.OrdinalIgnoreCase));
```

File: Src/FluentAssertions/Common/TypeExtensions.cs [hunk 4]
```diff
@@ -201,19 +194,12 @@ public static PropertyInfo FindProperty(this Type type, string propertyName)
...
-                return field;
-            }
-
-            type = type.BaseType;
-        }
+        var fields = type.GetNonPrivateFields(visibility);
 
-        return null;
+        return fields.FirstOrDefault(p =>
+            p.Name == fieldName || p.Name.EndsWith("." + fieldName, StringComparison.OrdinalIgnoreCase));
```
Comment (jnyrup) [inline Src/FluentAssertions/Common/TypeExtensions.cs:202]:
```suggestion
        return Array.Find(fields, p =>
            p.Name == fieldName || p.Name.EndsWith("." + fieldName, StringComparison.OrdinalIgnoreCase));
```

File: Src/FluentAssertions/Common/TypeExtensions.cs (hunk 3 above)
Comment (Corniel) [reply Src/FluentAssertions/Common/TypeExtensions.cs:188]: I would even suggest to define aan extension Method for this, so that you ca do: properties.Find()

Comment (dennisdoomen) [reply Src/FluentAssertions/Common/TypeExtensions.cs]: Nah, I would only do that when there's multiple uses and it would significantly improve readability.

Comment (Corniel) [reply Src/FluentAssertions/Common/TypeExtensions.cs]: @dennisdoomen In some of my repos, I had multiple places where I did this. If that is not the case here, I fully agree.

File: Src/FluentAssertions/Common/TypeMemberReflector.cs [hunk 5]
```diff
@@ -31,27 +31,45 @@ private static PropertyInfo[] LoadNonPrivateProperties(Type typeToReflect, Membe
...
-                .Where(property => includeInternals || property.GetMethod is { IsAssembly: false, IsFamilyOrAssembly: false })
+                .Where(property => includeInternal || !IsInternal(property))
+                .Select(property => new
+                {
+                    Property = property,
+                    IsExplicit = IsExplicitImplementation(property)
+                })
+                .OrderBy(x => x.IsExplicit)
+                .Select(x => x.Property)
                 .ToArray();
```
[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Src/FluentAssertions/Common/TypeMemberReflector.cs:57]: `IsExplicit` is a bool and initialized by the `Select`. So what's the problem here?

[the remark this answers is not in the document]
Comment (jn-foreflight) [reply Src/FluentAssertions/Common/TypeMemberReflector.cs:57]: There's no problems in terms of corrects, but we can remove two `Select`s and still only invoke `IsExplicitImplementation` `n` times.

[the remark this answers is not in the document]
Comment (jnyrup) [reply Src/FluentAssertions/Common/TypeMemberReflector.cs:57]: There's no problems in terms of correctness, but we can remove the two `Select`s and still only invoke `IsExplicitImplementation` `n` times.

[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Src/FluentAssertions/Common/TypeMemberReflector.cs]: Fixed it.

File: Src/FluentAssertions/Common/TypeMemberReflector.cs [hunk 6]
```diff
@@ -31,27 +31,39 @@ private static PropertyInfo[] LoadNonPrivateProperties(Type typeToReflect, Membe
...
+    private static IEnumerable<PropertyInfo> GetPropertiesFromHierarchy(Type typeToReflect, MemberVisibility memberVisibility)
     {
-        bool includeInternals = memberVisibility.HasFlag(MemberVisibility.Internal);
+        bool includeInternal = memberVisibility.HasFlag(MemberVisibility.Internal);
 
         return GetMembersFromHierarchy(typeToReflect, type =>
         {
             return type
                 .GetProperties(AllInstanceMembersFlag | BindingFlags.DeclaredOnly)
-                .Where(property => property.GetMethod?.IsPrivate == false)
```
Comment (jnyrup) [inline Src/FluentAssertions/Common/TypeMemberReflector.cs:49]: This also filter out properties without a getter.
How about we change it to just check if `property.GetMethod` is not null?

File: Src/FluentAssertions/Common/TypeMemberReflector.cs [hunk 7]
```diff
@@ -31,27 +31,39 @@ private static PropertyInfo[] LoadNonPrivateProperties(Type typeToReflect, Membe
...
+    {
+        return property.GetMethod is { IsAssembly: true } or { IsFamilyOrAssembly: true };
+    }
+
+    private static bool IsExplicitImplementation(PropertyInfo property)
+    {
+        return property.GetMethod?.IsPrivate == true &&
+            property.SetMethod?.IsPrivate == true &&
+            property.Name.Contains('.', StringComparison.Ordinal);
+    }
```
Comment (jnyrup) [inline Src/FluentAssertions/Common/TypeMemberReflector.cs:65]: Needs to be `property.SetMethod?.IsPrivate != false` to account for explicitly implemented properties without a setter.

If we're okay with letting this method know that we're filtered out properties without getters earlier (the commend above), we can remove the `?` from `GetMethod`

**Review by jnyrup — COMMENTED** [review]
Here a my findings so far - I'll stop thinking for today.
I'm wondering if we should move this PR to V7 as it is changing quite a bit on the internals.
It would be

File: Src/FluentAssertions/Equivalency/Matching/TryMatchByNameRule.cs [hunk 8]
```diff
@@ -10,14 +10,17 @@ internal class TryMatchByNameRule : IMemberMatchingRule
 {
     public IMember Match(IMember expectedMember, object subject, INode parent, IEquivalencyAssertionOptions options)
     {
-        PropertyInfo property = subject.GetType().FindProperty(expectedMember.Name);
+        PropertyInfo property = subject.GetType().FindProperty(expectedMember.Name,
+            options.IncludedProperties | MemberVisibility.ExplicitlyImplemented);
```
[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Src/FluentAssertions/Equivalency/Matching/TryMatchByNameRule.cs:14]: Good point. Apparently we don't have any tests for that. Will add them.

File: Src/FluentAssertions/Common/TypeMemberReflector.cs [hunk 9]
```diff
@@ -123,10 +140,10 @@ private static List<FieldInfo> GetFieldsFromHierarchy(Type typeToReflect, Member
             members.InsertRange(0, newPropertyInfos);
         }
 
-        return members;
+        return members.ToArray();
```
[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Src/FluentAssertions/Common/TypeMemberReflector.cs:143]: Because Qodana wanted me to use a more abstract type than `List<TMemberINfo` as the return value of `GetInterfaceMembers`. I could try to change it to `IEnumerablt<T>`

File: Tests/FluentAssertions.Equivalency.Specs/SelectionRulesSpecs.cs [hunk 10]
```diff
@@ -1779,10 +1779,33 @@ public void When_respecting_declared_types_explicit_interface_member_on_interfac
             subject.VehicleId = 1; // interface member
 
             // Act
-            Action action = () => subject.Should().BeEquivalentTo(expected, opt => opt.RespectingDeclaredTypes());
+            Action action = () => subject.Should().BeEquivalentTo(expected);
 
             // Assert
-            action.Should().NotThrow();
+            action.Should().Throw<XunitException>();
```
[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Tests/FluentAssertions.Equivalency.Specs/SelectionRulesSpecs.cs:1785]: Because of the changed/improved behavior, the _normal_ property `VehicleId` will now be taking into account instead of the one that was explicitly implemented. And since the _normal_ `VehicleId` of the subject is 2, and `1` is expected, it _should_ throw.

**Comment by dennisdoomen** [comment]
> I'm wondering if we should move this PR to V7 as it is changing quite a bit on the internals.

Given that you've found some more edge cases for which we don't even have tests yet, I'm fine moving this to v7 so we can expedite the final release. As far as I'm concerned, it's your call.

File: Src/FluentAssertions/Common/TypeMemberReflector.cs (hunk 6 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Common/TypeMemberReflector.cs:49]: Are you saying that we _should_ filter out properties without a getter?

Comment (dennisdoomen) [reply Src/FluentAssertions/Common/TypeMemberReflector.cs:49]: What would be the added value?

File: Src/FluentAssertions/Common/TypeMemberReflector.cs (hunk 7 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Common/TypeMemberReflector.cs:65]: Added an extra test to cover this scenario.

File: Src/FluentAssertions/Common/TypeExtensions.cs [hunk 11]
```diff
@@ -202,34 +196,27 @@ public static PropertyInfo FindProperty(this Type type, string propertyName)
...
-                return field;
-            }
+        var fields = type.GetFields(memberVisibility);
 
-            type = type.BaseType;
-        }
-
-        return null;
+        return Array.Find(fields, p =>
+            p.Name == fieldName || p.Name.EndsWith("." + fieldName, StringComparison.OrdinalIgnoreCase));
```
Comment (IT-VBFK) [inline Src/FluentAssertions/Common/TypeExtensions.cs:204]: Should this also be `.OrdinalIgnoreCase` -> `.Ordinal`

**Comment by dennisdoomen** [comment]
@jnyrup would be nice if you could give this one some attention soon...🙏

**Comment by jnyrup** [comment]
I must admit I have postponed further reviewing this PR partly because it wants to solve multiple problems when using `BeEquivalentTo` together with explicitly implemented members and I've found it difficult to see what changes related to which problem.
Is it possible to split it further up?

E.g. `Can_find_explicitly_implemented_property_on_the_subject` from #2146 works if I rewrite `FindProperty` to

```cs
public static PropertyInfo FindProperty(this Type type, string propertyName)
{
    while (type != typeof(object))
    {
        PropertyInfo[] properties = type.GetProperties(AllInstanceMembersFlag | BindingFlags.DeclaredOnly);
        if (Array.Find(properties, e => e.Name == propertyName) is { } property)
        {
            return property;
        }

        if (Array.Find(properties, e => e.Name.EndsWith("." + propertyName, StringComparison.Ordinal)) is { } explicitProperty)
        {
            return explicitProperty;
        }

        type = type.BaseType;
    }

    return null;
}
```

File: Src/FluentAssertions/Common/TypeMemberReflector.cs (hunk 6 above)
Comment (jnyrup) [reply Src/FluentAssertions/Common/TypeMemberReflector.cs:49]: If `includeInternal` is true, we can now return a member without a getter, whereas currently `Where(property => property.GetMethod?.IsPrivate == false)` requires the property to have a getter.

But I see this doesn't matter in practice since `HasGetter` will discard any property without a getter .
image

So I withdraw my comments on this one.

File: Src/FluentAssertions/Common/TypeMemberReflector.cs [hunk 12]
```diff
@@ -145,12 +162,18 @@ private static List<FieldInfo> GetFieldsFromHierarchy(Type typeToReflect, Member
             typeToReflect = typeToReflect.BaseType;
         }
 
-        return members;
+        return members.ToArray();
```
Comment (jnyrup) [inline Src/FluentAssertions/Common/TypeMemberReflector.cs:165]: 🙃 Doesn't seem necessary

File: docs/_pages/releases.md [hunk 13]
```diff
@@ -46,6 +46,7 @@ sidebar:
 
 ### Fixes
 * `because` and `becauseArgs` were not included in the error message when collections of enums were not equivalent - [#2214](https://github.com/fluentassertions/fluentassertions/pull/2214)
+* `BeEquivalentTo` will now find and can map subject properties that are implemented through an explicitly-implemented interface - [2152](https://github.com/fluentassertions/fluentassertions/pull/2152)
```
Comment (jnyrup) [inline docs/_pages/releases.md:49]:
```suggestion
* `BeEquivalentTo` will now find and can map subject properties that are implemented through an explicitly-implemented interface - [#2152](https://github.com/fluentassertions/fluentassertions/pull/2152)
```

**Review by jnyrup — APPROVED** [review]
